sdkbuild/typescript.go

```diff
@@ -0,0 +1,201 @@
+package sdkbuild
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+// BuildTypeScriptProgramOptions are options for BuildTypeScriptProgram.
+type BuildTypeScriptProgramOptions struct {
+	// Directory that will have a temporary directory created underneath.
+	BaseDir string
+	// Required version. If it contains a "/" it is assumed to be a path with a
+	// package.json. Otherwise it is a specific version (with leading "v" is
+	// trimmed if present).
+	Version string
+	// Required set of paths to include in tsconfig.json paths for the harness.
+	// The paths should be relative to one-directory beneath BaseDir.
+	HarnessPaths map[string][]string
+	// If present, this directory is expected to exist beneath base dir. Otherwise
+	// a temporary dir is created.
+	DirName string
+	// If present, applied to build commands before run. May be called multiple
+	// times for a single build.
+	ApplyToCommand func(context.Context, *exec.Cmd) error
+}
+
+// TypeScriptProgram is a TypeScript-specific implementation of Program.
+type TypeScriptProgram struct {
+	dir string
+}
+
+var _ Program = (*TypeScriptProgram)(nil)
+
+// BuildTypeScriptProgram builds a TypeScript program. If completed
+// successfully, this can be stored and re-obtained via
+// TypeScriptProgramFromDir() with the Dir() value (but the entire BaseDir must
+// be present too).
+func BuildTypeScriptProgram(ctx context.Context, options BuildTypeScriptProgramOptions) (*TypeScriptProgram, error) {
+	if options.BaseDir == "" {
+		return nil, fmt.Errorf("base dir required")
+	} else if options.Version == "" {
+		return nil, fmt.Errorf("version required")
+	} else if len(options.HarnessPaths) == 0 {
+		return nil, fmt.Errorf("at least one harness path required")
+	}
+
+	// Create temp dir if needed that we will remove if creating is unsuccessful
+	success := false
+	var dir string
+	if options.DirName != "" {
+		dir = filepath.Join(options.BaseDir, options.DirName)
+	} else {
+		var err error
+		dir, err = os.MkdirTemp(options.BaseDir, "program-")
+		if err != nil {
+			return nil, fmt.Errorf("failed making temp dir: %w", err)
+		}
+		defer func() {
+			if !success {
+				// Intentionally swallow error
+				_ = os.RemoveAll(dir)
+			}
+		}()
+	}
+
+	// Create package JSON
+	var packageJSONDepStr string
+	if strings.Contains(options.Version, "/") {
+		if _, err := os.Stat(filepath.Join(options.Version, "package.json")); err != nil {
+			return nil, fmt.Errorf("failed finding package.json in version dir: %w", err)
+		}
+		localPath := "file:" + options.Version
+		pkgs := []string{"activity", "client", "common", "internal-workflow-common",
+			"internal-non-workflow-common", "proto", "worker", "workflow"}
+		for _, pkg := range pkgs {
+			packageJSONDepStr += fmt.Sprintf(`"@temporalio/%v": "%v/packages/%v",`, pkg, localPath, pkg)
+			packageJSONDepStr += "\n    "
+		}
```

Comment:
I really think we should keep the template. It's eminently more readable than this (and already was set up). This feels like a step backwards.

The others probably deserve to be as well, but we don't need to do that right now.

Reply:
Templates do not travel with the library as a dependency and templates aren't how the other files for this repo were done, they were done with strings as well (go.mod, main.go, build.gradle, settings.gradle, pyproject.toml, etc). This returns to consistency with those (unless we should turn everything to templates and then embed templates into the library for distribution, but not sure that is worth it during this refactor effort).

Reply:
Let's just embed them then, seems easy enough. I think it's quite a bit more readable - especially this particular bit I highlighted.

Reply:
Changed logic anyways to work on Windows (properly escape backslashes in JSON strings properly) and to work with relative paths to cloned SDK directories.